### ISSUE TITLE
We could create new CantusIdentifier model

### ISSUE THREAD

**Issue by jacobdgm** [DDMAL/CantusDB#410]
I suggest that we add a new model named `CantusIdentifier`. This new model would address two issues:

First, "a way to check that all the genres in the Cantus Database are identical to the Cantus Index Cantus ID # entries" has been requested in the wish list (further: "Sometimes errors slip in and human eyes don't see that the genre has been entered incorrectly. (I.e., Can ID 202838 should always have the genre "A". Each Can ID # should have only 1 genre assigned with it, and tell me if you find any that do not match so I can fix them).") This type of information could conveniently be stored in a CantusIdentifier object.

Second, it would help optimize some things related to the local storage of concordances retrieved from Cantus Index etc. (as discussed in #405). For some Cantus IDs, we have dozens of chants associated with that ID. By having a single object associated with each individual ID, we can store the concordance only once for each ID (rather than dozens of times; this is only a small concern, though, because it doesn't take a lot of data to store concordances), and when we update our stored concordances, we only need to send a single request to Cantus Index (rather than dozens).

**Comment by jacobdgm** [comment]
In a video call yesterday, Debra noted a peculiarity with the way that Cantus Index works - if we created a `CantusIdentifier` model, we could help Cantus Index run more smoothly.

As far as we understand, in CI, the concordances information stored there is only updated when a user like Debra presses an "update" button somewhere on the site, or when a request is made to a `json-con/<cantus id>/refresh` API.

As I suggested above, we could keep information on each Cantus ID in our own database. We could update this information (say) once a week with a `cron` job, sending a request to the `/refresh` API so that we have the most up-to-date information possible. If we did this, we would/could:
- be able to fetch data on Cantus IDs from our own database, improving page load times
- ensure that the Cantus Index's is updated at least once a week

I told Debra that she shouldn't expect this to be implemented any time soon, so it remains low priority. But I think it could be worth looking into at some point in the future, since it has potential benefits both for NewCantus and for Cantus Index.

**Comment by jacobdgm** [comment]
An email I just got from Jan re: #1029 has got me thinking about this again. The potential effects of creating this include:
- CantusIDs would be a controlled vocabulary - when creating/editing chants, users would have to select from a list of existing CantusIDs instead of entering one freehand (thus limiting the possibility of typos)
  - we can also ensure that when a user selects a Cantus ID, the chant's Genre is automatically set to the genre associated with the CantusID
  - we would need to talk with Debra about how new CantusIDs are created - if a user is indexing a manuscript, will they ever need to invent a new CantusID? Would such a CantusID need to be added to CantusIndex first before it can be selected in CantusDB
- It would be easy/efficient to cache concordances from CantusIndex, since these concordances belong to specific CantusIDs and not to specific chants. This would also allow us to continue to display concordances even in case of a temporary outage on CantusIndex
- For the Input Tool, this would allow us to store canonical texts and genres for all CantusIDs
- For the previous two points, performance would probably also be improved, because we won't need to send requests to CantusIndex in order to load our pages - we'd just need to query our own database.

**Comment by annamorphism** [comment]
>       * we can also ensure that when a user selects a Cantus ID, the chant's Genre is automatically set to the genre associated with the CantusID

Doesn't this happen already with the Input Tool? Or do you mean in some other places one might select a chant? (Some Cantus IDs do have multiple genres associated with them, so proceed with caution).

>       * we would need to talk with Debra about how new CantusIDs are created - if a user is indexing a manuscript, will they ever need to invent a new CantusID? Would such a CantusID need to be added to CantusIndex first before it can be selected in CantusDB

Yes, users do encounter the need to invent new IDs some of the time. The workflow here is typically for the indexer to leave it blank, somebody higher up (me, Debra, the supervisor of their research project...) to enter it into Cantus Index, and then insert the newly created ID into the blank. What you're proposing doesn't seem to change this much.

**Comment by jacobdgm** [comment]
> Doesn't this happen already with the Input Tool? Or do you mean in some other places one might select a chant? (Some Cantus IDs do have multiple genres associated with them, so proceed with caution).

Yes, this happens with the Input Tool. I suggested this in response to Debra's comment:

> Sometimes errors slip in and human eyes don't see that the genre has been entered incorrectly. (I.e., Can ID 202838 should always have the genre "A". Each Can ID # should have only 1 genre assigned with it, and tell me if you find any that do not match so I can fix them)

So if some Cantus IDs have multiple associated genres, this contradicts what Debra said. In any case, Debra didn't ask for this to be enforced on, say, the database level, so it would be fine to not implement a feature that checks for this on the Chant Create/Edit pages. I'll update my most recent post. But a CantusIdentifier model would make it (slightly) simpler to answer questions about chants associated with particular Cantus IDs and particular Genres.

**Comment by jacobdgm** [comment]
raising priority from low to high. Setting up this model will allow us to more gently and robustly obtain and store concordances information from CI.

**Comment by jacobdgm** [comment]
If my understanding is the same as that of others, I believe we've decided not to add such a new model to the database. If we decide we need to implement a controlled vocabulary, we'll do it with a list of acceptable Cantus IDs, and if we need to get the expected fulltext for a Cantus ID, we'll get it from a cached copy of an API call to Cantus Index. Closing.
